### PR TITLE
Rule fixes based on QA report #260

### PR DESCRIPTION
**Pull request by abhishekmadan30**

### CODE REVIEW

**Comment by cadeyrn** [comment]
@abhishekmadan30: Why do you remove rules in your PR that were recently added?

File: cookie-banner-rules-list.json [hunk 1]
```diff
@@ -540,34 +537,31 @@
...
       "domains": ["netflix.com"]
     },
     {
       "click": {
-        "optIn": "div[role=dialog] > div > button",
-        "optOut": "div[role=dialog] > div > button + button",
-        "presence": "#scrollview + div"
+        "optIn": "._a9--._a9_0",
+        "optOut": "._a9--._a9_1",
+        "presence": ".x1n2onr6.xzkaem6"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:552]: This is slightly better than the old rule (which could have targeted any dialog with that structure). But I'm still worried that this isn't specific enough to the cookie banner. Also it may break when their front-end is redeployed.
I suggest that we try to find a cookie rule instead. If that doesn't work let's remove the rule entirely.

File: cookie-banner-rules-list.json [hunk 2]
```diff
@@ -540,34 +537,31 @@
...
-      "click": {},
-      "cookies": {
-        "optOut": [
-          { "name": "ckns_explicit", "value": "2" },
-          { "name": "ckns_policy", "value": "000" },
-          { "name": "ckns_privacy", "value": "july2019" }
-        ]
+      "click": {
+        "presence": "#bbccookies",
+        "optIn": "#bbccookies-continue-button"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:562]: The banner has a "do not agree" button, can we target that too?

image

Oddly this site has two banners. Since we can't handle both let's target the full screen one instead, which is most disrupting.

File: cookie-banner-rules-list.json [hunk 3]
```diff
@@ -580,7 +574,7 @@
         ]
       },
       "id": "290db348-ced2-4f16-9954-da476d0aef01",
-      "domains": ["web.de", "gmx.net"]
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:583]: Please revert this change! I assume this is a merge gone wrong. 🙂

File: cookie-banner-rules-list.json [hunk 4]
```diff
@@ -665,13 +659,11 @@
...
     {
-      "click": {},
-      "cookies": {
-        "optOut": [
-          { "name": "euconsent-v2", "value": "1" },
-          { "name": "euconsent-v2-noniab", "value": "AAYA" }
-        ]
+      "click": {
+        "optIn": ".components-button.white-space-normal.is-primary",
+        "presence": ".components-modal__screen-overlay"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:664]: This is too generic, can you find a click rule that is more specific to the cookie banner?

File: cookie-banner-rules-list.json [hunk 5]
```diff
@@ -1167,7 +1167,7 @@
         "optOut": [
           {
             "name": "FCCDCF",
-            "value": "%5Bnull%2Cnull%2Cnull%2C%5B%22CPgbQkAPgbQkAEsABCBGCjCgAAAAAAAAAAIwAAAOhQD2F2K2kKEkfjSUWYAQBCujIEIBUAAAAECBIAAAAUgQAgFIIAgAAlACAAAAABAQAQCAgAQABAAAoACgACAAAAAAAAAAAAQQAABAAIAAAAAAAAEAQAAAAAQAAAAAAABEhCAAQQAEAAAAAAAAAAAAAAAAAAABAAA.YAAAAAAAAAA%22%2C%221~%22%2C%22FA11CE3A-809E-4B87-BFDE-13E9DE22F419%22%5D%2Cnull%2Cnull%2C%5B%5D%5D"
+            "value": "%5Bnull%2Cnull%2Cnull%2C%5B%22CP00b0AP00b0AEsACCBGDeCgAAAAAAFAAAIwAAAQaQD2F2K2kKFgPi2QWYAQBCij4EAhQAAAAkCBIAAgAUgQAgFIIAwAIFAAAAAAAAAQEgCQAAQABAAAIACgAAAAAAIAAAAAAAQQAAAAAIAAAAAAAAEAQCAAAAQAAAAIAABEhCAAQQAEAAAAAAAQAAAAAAAAAAABAAAAAAAAAAAAAAAAAgAAgAA.YAAAAAAAAAA%22%2C%221~%22%2C%22905C9CAE-3D41-4084-95DA-144E5455EA05%22%5D%2Cnull%2Cnull%2C%5B%5D%5D"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:1170]: Are you sure these cookies will work in the long term? They contain IDs. Please either add a click rule or remove the rule if that doesn't work.

File: cookie-banner-rules-list.json [hunk 6]
```diff
@@ -1368,7 +1354,7 @@
         "optOut": [
           {
             "name": "FCCDCF",
-            "value": "%5Bnull%2Cnull%2Cnull%2C%5B%22CPgejgAPgejgAEsABCBGCkCgAAAAAAAAAAIwAAAOhQD2F2K2kKEkfjSUWYAQBCujIEIhUAAAAECBIAAAAUgQAgFIIAgAAlACAAAAABAQAQCAgAQABAAAoACgAAAAAAAAAAAAAAQQAABAAIAAAAAAAAEAQAAIAAQAAAAAAABEhCAAQQAEAAAAAAAAAAAAAAAAAAABAAA.YAAAAAAAAAA%22%2C%221~%22%2C%22B14550D2-2C81-4087-B1D1-A35C539B751B%22%5D%2Cnull%2Cnull%2C%5B%5D%5D"
+            "value": "%5Bnull%2Cnull%2Cnull%2C%5B%22CP00b0AP00b0AEsACCBGDeCgAAAAAEPAAAIwAAAQaQD2F2K2kKFgPi2QWYAQBCij4EAhQAAAAkCBIAAgAUgQAgFIIAwAIFAAAAAAAAAQEgCQAAQABAAAIACgAAAAAAIAAAAAAAQQAAAAAIAAAAAAAAEAQCAAAAQAAAAIAABEhCAAQQAEAAAAAAAQAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAgAA.YAAAAAAAAAA%22%2C%221~%22%2C%229CD0A99E-8745-4688-A861-B29CD9A5460F%22%5D%2Cnull%2Cnull%2C%5B%5D%5D"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:1357]: Same as above, it's unlikely that this rule will work long term.

File: cookie-banner-rules-list.json [hunk 7]
```diff
@@ -1534,23 +1511,11 @@
...
-          }
-        ],
-        "optOut": [
-          {
-            "name": "rbzid",
-            "value": "9B4wIoP8rsa+1YSv57XFo/uUSV+sZUNn9YHPdzFdQ9vj99CJquEXFnM9Icin8eAg4s61utnrutr/pbB4VnjxhdpAYo9E4+hvwG1ldnRSNgrgH67CESdXEK8ui4aWrxYINvbBwjSgrFpKg31SfphpXBJ+Rn2DUGX8HsLFm6gd0Erqfm1LGPuP0KbsVo+XacG5fDA4pTQ8CorfGyH8TTH9hIe9NozfxGveJAX7VoPFDpH7+Vmj9DlOUQ4kVR80UIhm"
-          }
-        ]
+        "optIn": ".message-component.message-button.no-children.focusable.primary.sp_choice_type_11.last-focusable-el",
+        "presence": ".message-container",
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:1515]: This looks a lot like the other sourcepoint rules (look for ".sp_choice_type"). Can it be merged into those rules?

File: cookie-banner-rules-list.json [hunk 8]
```diff
@@ -1639,7 +1587,18 @@
...
+      "click": {
+        "optIn": "#didomi-notice-agree-button",
+        "optOut": ".didomi-continue-without-agreeing",
+        "presence": "div#didomi-notice"
+      },
+      "cookies": {},
+      "id": "688d29a8-e1c7-4d62-b3d4-53b451ff5a48",
+      "domains": [
         "orange.sk"
       ]
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:1603]: There are already rules for didomi, please check if this rule can be merged into those.

File: cookie-banner-rules-list.json [hunk 9]
```diff
@@ -2118,18 +2063,21 @@
...
             "name": "wscrCookieConsentShared",
-            "value": "1=true&2=true&3=true&4=true&5=true&visitor=03bfbe89-44d7-4da3-ad3f-a1a8a8285257&version=20221005-001"
+            "value":"1=true&2=true&3=true&4=true&5=true&visitor=44dd8519-4cd6-46ab-bf7d-7988c112da26&version=20231107-001"
           }
         ],
         "optOut": [
           {
             "name": "wscrCookieConsentShared",
-            "value": "1=true&2=false&3=false&4=false&5=false&visitor=5698152c-e2fd-4076-b411-f8baed432e40&version=20221005-001"
+            "value": "1=true&2=false&3=false&4=false&5=false&visitor=44dd8519-4cd6-46ab-bf7d-7988c112da26&version=20231107-001"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:2080]: We shouldn't inject cookies with uuids since they are likely to be unstable. Let's rely only on clicking instead.

File: cookie-banner-rules-list.json [hunk 10]
```diff
@@ -2186,13 +2134,13 @@
...
+            "value": "%5B%5B%22AKsRol8XFyL-33asXcbtHS5bfjDNSv1R-S70LyCQmv78L4aowSzqaRNzkbzRDoSbM_e3lki6ntIZT_LPzrp51sfs1tqi5wtb7Dd6_5euFHvTjhfP6-gJK2x6bEUqKswj5u79VR8NwnAHNQ6XEK3x2pFpDY2ybDzqrw%3D%3D%22%5D%2Cnull%2C%5B%5D%5D"
           }
         ],
         "optOut": [
           {
             "name": "FCCDCF",
-            "value": "%5Bnull%2Cnull%2Cnull%2C%5B%22CPgvCMAPgvCMAEsABCHRCkCgAAAAAAAAAA6IAAARTY7-N262v7O8fnfUebo0Rq-r-Uv1cBmHCOLRo4BOX8t-VgH-sJ2ALt0aVgRMjxJZqQCQgISAJOICpNi0RIqWiGQQAkxRAaQQ2DJILYy8sCgIJIfFxKCsAUzy9zcSM7zklvx3Tf9v7lut7rrfiZet8HLoXGz9xbaNtbe8AIywcE4ZqSVrxaZ-d__pmcAA.YAAAAAAAAAA%22%2C%221~%22%2C%228B9BF72F-C252-4FA2-A91E-8632E50EC476%22%5D%2Cnull%2Cnull%2C%5B%5D%5D"
+            "value": "%5Bnull%2Cnull%2Cnull%2C%5B%22CP07BsAP07BsAEsACCHRDeCgAAAAAH_AAA6IAAASMYr-N262v7P4Pi_Q-boURqyr-En1YBiACuDRowBuXUt-RgHusJ2ALN0YVwQMjAJRugCQAISABOICLNikRAqXiGYAIoxQkaQQcDIAOYi8MCgAJAfFxLCkAUzy9TY6MzzklvxyTftvrlms6pPfmZat8GL4fGw9waKNtbesAIjxeQ47qyVrxaZ9Z_SJgUAACAAAAAAgAA.YAAAD_gAAAA%22%2C%221~%22%2C%223DC34A0E-C6A7-4308-A161-08B2D340ECA0%22%5D%2Cnull%2Cnull%2C%5B%5D%5D"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:2143]: Same as above, IDs may be user or session specific and therefore unstable.

File: cookie-banner-rules-list.json [hunk 11]
```diff
@@ -2742,6 +2664,18 @@
...
+        "optIn": "#didomi-notice-agree-button",
+        "optOut": ".didomi-continue-without-agreeing",
+        "presence": "#didomi-notice"
+      },
+      "cookies": {},
+      "id": "ecd2364d-8f24-41b9-93a8-5355951aeb4b",
+      "domains": [
+        "hnonline.sk"
+      ]
+    },
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:2678]: Can this be merged into the other didomi rules?

File: cookie-banner-rules-list.json [hunk 12]
```diff
@@ -2855,9 +2789,9 @@
     },
     {
       "click": {
-        "optIn": "button[data-testid=\"cookie-banner-strict-accept-all\"]",
-        "optOut": "button[data-testid=\"cookie-banner-strict-accept-selected\"]",
-        "presence": "div[data-testid=\"dl-cookieBanner\"]"
+        "optIn": ".text-deepl-blue active:bg-dark-7 rounded-xs flex items-center justify-center bg-white px-4 py-1 font-medium",
+        "optOut": ".active:border-dark-7 active:text-dark-7 rounded-xs flex items-center justify-center border border-white px-4 py-1 font-medium text-current",
+        "presence": ".focus-visible-light-container bg-deepl-blue text-white sm:rounded-lg"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:2794]: These selectors rely on styling classes rather than stable identifiers that are cookie banner specific. Are the old selectors not working anymore? Looking at the sites DOM the attributes they use are still there. Also `document.querySelector` still returns the buttons properly. Please double check!

File: cookie-banner-rules-list.json [hunk 13]
```diff
@@ -3168,9 +3092,8 @@
...
     {
       "click": {
-        "optIn": "button.fc-cta-consent",
-        "optOut": "button.fc-cta-do-not-consent",
-        "presence": "div.fc-dialog-container"
+        "optIn": ".fc-button.fc-cta-consent.fc-primary-button",
+        "presence": ".fc-dialog-container"
       },
       "cookies": {},
       "id": "ff596e06-7834-4f75-b2a7-ccb39fad8925",
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:3099]: Can this rule be merged with `5fd67d61-aaa7-4431-af6f-0f1c31c849fc` ?

File: cookie-banner-rules-list.json [hunk 14]
```diff
@@ -3744,9 +3668,9 @@
     },
     {
       "click": {
-        "optIn": "button.corgi__sc-140l2f3-1",
-        "optOut": "button.corgi__sc-1k5bfrh-3",
-        "presence": "div.bJyFEZ"
+        "optIn": ".corgi__sc-140l2f3-0.corgi__sc-140l2f3-1.corgi__sc-1k5bfrh-2.foVfGp.gjMFpW.esJNpe",
+        "optOut": ".corgi__sc-140l2f3-0.corgi__sc-1k5bfrh-3.gIWeVI.pjCEP",
+        "presence": ".corgi__sc-p6yu06-0.UAcHO"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:3673]: Instead of relying on these semi-random class names, can we use the `data-testid` attributes instead? For exampel the reject-all button has `data-testid="cookie-wall-reject"`.

File: cookie-banner-rules-list.json [hunk 15]
```diff
@@ -3791,8 +3715,9 @@
     },
     {
       "click": {
-        "optIn": "span#hf_cookie_text_cookieAccept",
-        "presence": "div.pre-cookie-modal-body"
+        "optIn": ".nds-btn.dialog-actions-accept-btn.css-60b779.ex41m6f0.btn-primary-dark.btn-md",
+        "optOut": ".nds-btn.dialog-actions-decline-btn.css-60b779.ex41m6f0.btn-primary-dark.btn-md",
+        "presence": ".dialog-content-wrapper.dialog-appear-done.dialog-enter-done"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:3720]: Please don't use styling classes for the selector as it may break when the style changes. Would'nt e.g. `.dialog-actions-accept-btn` be enough?

File: cookie-banner-rules-list.json [hunk 16]
```diff
@@ -5382,9 +5316,9 @@
     },
     {
       "click": {
-        "presence": "#mms-consent-portal-container",
-        "optOut": "button[data-test=\"pwa-consent-layer-deny-all\"]",
-        "optIn": "button[data-test=\"pwa-consent-layer-accept-all\"]"
+        "presence": "#mms-consent-portal-container, .cmpbox.cmpstyleroot.cmpbox3.cmpboxWelcomeGDPR.cmpBoxWelcomeOI",
+        "optOut": "button[data-test=\"pwa-consent-layer-deny-all\"], .cmpboxbtn.cmpboxbtnno.cmptxt_btn_no",
+        "optIn": "button[data-test=\"pwa-consent-layer-accept-all\"], .cmpboxbtn.cmpboxbtnyes.cmptxt_btn_yes"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:5321]: Why do we need multiple selectors here?

File: cookie-banner-rules-list.json [hunk 17]
```diff
@@ -5462,27 +5396,21 @@
...
     {
-      "id": "a8222cf2-8f6c-48df-8215-05a15d4ac592",
-      "domains": ["hetzner.com"],
-      "cookies": {
-        "optOut": [
-          {
-            "name": "CookieConsentDeclined",
-            "value": "1"
-          }
-        ]
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:5485]: Please don't remove this rule, it was just added.

File: cookie-banner-rules-list.json [hunk 18]
```diff
@@ -5462,27 +5396,21 @@
...
     {
-      "id": "dea34d82-9c05-4c08-9262-18a7f62be91e",
-      "domains": ["netcup.de"],
-      "cookies": {
-        "optOut": [
-          {
-            "name": "js-cookie-opt-in__consent",
-            "value": "needed"
-          }
-        ]
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:5473]: Please don't remove this rule, it was just added.

**Review by Trikolon — CHANGES_REQUESTED** [review]
Thanks for updating so many rules! I've found a few issues for which I've added comments inline.

Please also run `npm run test` locally and make sure it passes. This checks for both rule validity and formatting issues.

File: cookie-banner-rules-list.json (hunk 2 above)
Comment (abhishekmadan30) [reply cookie-banner-rules-list.json:562]: I don't get the full screen banner.

File: cookie-banner-rules-list.json (hunk 4 above)
Comment (abhishekmadan30) [reply cookie-banner-rules-list.json:664]: Outside of parsing the text within the button, there is nothing more specific to banner

Comment (Trikolon) [reply cookie-banner-rules-list.json:650]: In that case we should remove the rule. We shouldn't risk clicking the wrong thing.
The class `.cmp__dialog` seems useful, but the buttons only differ in text and role. Targeting based on role is too weak IMO.

File: cookie-banner-rules-list.json [hunk 19]
```diff
@@ -3715,9 +3629,9 @@
     },
     {
       "click": {
-        "optIn": ".nds-btn.dialog-actions-accept-btn.css-60b779.ex41m6f0.btn-primary-dark.btn-md",
-        "optOut": ".nds-btn.dialog-actions-decline-btn.css-60b779.ex41m6f0.btn-primary-dark.btn-md",
-        "presence": ".dialog-content-wrapper.dialog-appear-done.dialog-enter-done"
+        "optIn": ".dialog-actions-accept-btn",
+        "optOut": ".dialog-actions-decline-btn",
+        "presence": ".dialog-content-wrapper"
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:3634]: Too generic, this looks like it targets any dialog. I can see a `cookie-dialog-base` class, can we target that? It's fine if the buttons are more generic as long as the presence selector is specific. That's because we only query the buttons as children of the presence dialog.

**Review by Trikolon — CHANGES_REQUESTED** [review]
Thanks! Please mark comments you addressed as "resolved" so it's easier to see what work is still pending.

**Comment by Trikolon** [comment]
@abhishekmadan30 make sure you rebase on the latest version of `main`. This still seems to have conflicts as per the message above.

File: cookie-banner-rules-list.json [hunk 20]
```diff
@@ -4416,13 +4261,18 @@
...
       },
       "cookies": {
         "optOut": [
           {
             "name": "osano_consentmanager",
-            "value": "YsI6iYzib6nSBOK6D61-a6M0e1Cgk0B-MFLyyCSPPQ-GfL4knxA8syVY4wEmm3isBY6aDBQm2FGawF-2ou-zYT-hd5u78BRUjnQWaBnz9067nHpfBEIAnrgNH_unK6ew1HdgSLm-Y3_zOPFRQjVqFiC9Bdw3bcfqkdAJ6LzJ7arDUdh5ElFZpnBk5JsAeech5COxlZnJ8niJ4_4T3KO99S87PK3GK-SnwCvGHM02U9xoD3TeMNCTvi87ToGdcqfHEnCeLaExAXVrX7_RRXjc_UBOFmw="
+            "value": "L4Iw2hjgGqOMf2loNomB_36ZyLsqV8mAt_uNR24kLwoYU_P5toPrdi0UDvh3PeeovKUuw_Qu255Sd4SSh3gxIXo7zpAlBMnP3LSIUqUMQL5AeLtHmVXHvikZjI4je6_uN3VEELj3QdbPQp5AGEqOB5EcBNMZHA2Ikzodk6jMpTnP2eZIWDu8PWOJmdBK4KK3YI0GDvIRTq43yaZ5xtBOPY7U1xADPGNDk4s0dHdtE-0XSaRSndnVNagdGdnLfFLeKLfSG5Ks_btmdqHfCD1J3m06tVQGuR3TTKXM_rJGn1njzczdoF_qgVSh15XlJQPP4dmpKoFPUXxU_qngz1legDWJAKWOjw-EEy14twNoHmswqE8q3qiH_tN9HSfbBarSus1wswfR4xOEwREqZamf0T4HThBE0FZelh9cuh8DUcYUWm-cuVOYCxsN_2bLmrV0MbtnE2Z-hUXZxJgNYtk9gYHIW1xOZG6lfKH5C6SbppoHJNl2l_oqg4Wx6AYxRRiHWeDlyCXhMF7hQYZuzoLO5FgfacTOuA8E6phGD1ls3lM0dWVz"
+          },
+          {
+            "name": "osano_consentmanager_uuid",
```
Comment (Trikolon) [inline cookie-banner-rules-list.json:4274]: Please don't add UUIDs. Can we remove the cookie rule instead?

**Review by Trikolon — CHANGES_REQUESTED** [review]
I found one more issue, otherwise the changes look good. Thanks!

**Review by Trikolon — APPROVED** [review]
Thanks!
